# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0303 Торопыгин Антон, вариант 25.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.


## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать

## Основные теоретические положения

$\quad$ Аддитивная цепочка - последовательность натуральных чисел от 1 до n
$$ 1 = a_0, a_1, ..., a_r = n ,$$ 
где каждый элемент равен сумме каких-либо двух предыдущих элементов
$$ a_{i} = a_{j} + a_{k}, k \leq j < i $$

$ l(n) $ - наименьшая длина аддитивной цепочки для заданного $ n \in \mathbb{N} $.

Неравенство для метода множителей: $ \quad l(mn) \leq l(m) + l(n) $

Неравенство для m-арного метода $ [m = 2^k, n = \sum_{j=0}^{k} d_{j}m^{t-j}] $: $ \quad l(n) = m - 2 + (k+1)t $

Неравенство для бинарного алгоритма "SX": $ \quad l(n) = \lambda(n) + \nu(n) - 1 $

### Свойства аддитивных цепочек
* Полагается, что все аддитивные цепочки - возрастающие:
$ 1 = a_0 < a_1 < ... < a_r = n $
* Если два числа из последовательности $\{a_i\}$ совпадают, то одно из них может быть опущено
* Пара $ (j, k), 0 \leq k < j < i $ называется шагом
* Если существует более одной пары $ (j, k) $, полагается, что $ j $ - наибольшее из возможных 

### Виды шагов 
* удвоение: $ \quad j = k = i - 1 $
* звездный шаг: $ \quad j = i - 1 $
* малый шаг: $ \quad \lambda(a_{i}) = \lambda(a_{i-1}) $

### Свойства шагов
* Первый шаг - всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг
* За удвоением всегда следует звездный шаг
* Если $ i-й $ шаг не малый, то $ (i + 1)-й $ - либо звездный, либо малый либо и тот, и другой
* Если $ (i + 1)-й $ шаг ни звездный, ни малый, то $ i-й $ шаг должен быть малым

### Теорема
Если аддитивная цепочка включает $ d $ удвоений и $ f = r - d $ неудвоений, то 
$$ n \leq 2^{d-1}F_{f+3}, $$
где $ F_j $ - число Фибоначчи под номером $ j $.
### Следствие из теоремы
Если аддитивная цепочка содержит $ f $ удвоений и $ S $ малых шагов, то
$$ S \leq f \leq \frac{S}{1-log_2(\varphi)}, $$ 
где $ \varphi = \frac{\sqrt{5} + 1}{2} $ - золотое сечение.

### Алгоритм Брауэра
Для заданного числа $ n \in \mathbb{N} $ можно построить цепочку Брауэра с помощью рекуррентной формулы
$$ B_k(n) = 
\begin{cases} 
1, 2, 3, ..., 2^{k} - 1,\  если \  n < 2^{k}\\
B_k(q), 2q, ..., 2^kq, n,\  если \  n \geq 2^{k} и \  q = \lfloor{\frac{n}{2^k}} \rfloor
\end{cases} $$

Данная цепочка будет иметь длину
$$ l_B(n) = j(k + 1) + 2^k - 2 $$
при условии, что $ jk \leq \lambda(n) < (j + 1)k $

Длина цепочки будет минимизирована для больших $ n $, если 
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

#### Уточнения
* Задается фикированный параметр $ k $. Выполняются вычисления вспомогательных чисел:
$$ 
d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \ mod \ d \Rightarrow n = q_{1}d + r_{1} \ (0 \leq r_{1} < d) \\
q_2 = [\frac{q_1}{d}], r_2 = q_1 \ mod \ d \Rightarrow q_1 = q_{2}d + r_{2}
$$
* Данная процедура выполняется до тех пор, пока $ q_s $ не станет меньше $ d $. Следовательно $ q_{s-1} = q_{s}d + r_{s} $
* Таким образом,
$$ n = 2^kq_1 + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s)...) + r_2) + r_1 $$
$$ B_k(n): 1, 2, 3, ..., 2^k - 1, 2q_s, 4q_s, ..., 2^kq_s + r_s, ..., n $$

### Задание 1 

1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для 𝑥
𝑛 для 2-3 значений 𝑛 (значение 𝑛 > 30 выбираются студентом самостоятельно). Сравнить количество операций для
каждого метода, сделать выводы.


Возьмём числа: $$ a^{48} $$ $$ a^{63}$$
И применим к ним бинарный метод:

| № | N  | Y    | Z    |
|---|----|------|------|
| 0 | 48 | 1    | x    |
| 1 | 24 | 1    | x^2  |
| 2 | 12 | 1    | x^4  |
| 3 | 6  | 1    | x^8  |
| 4 | 3  | 1    | x^16 |
| 5 | 1  | x^16 | x^32 |
| 6 | 0  | x^48 | x^32 |

| № | N  | Y    | Z    |
|---|----|------|------|
| 0 | 63 | 1    | x    |
| 1 | 31 | x    | x^2  |
| 2 | 15 | x^3  | x^4  |
| 3 | 7  | x^7  | x^8  |
| 4 | 3  | x^15 | x^16 |
| 5 | 1  | x^31 | x^32 |
| 6 | 0  | x^63 | x^32 |

Количество операций, необходимых для возведения в степень вычисляется по формуле: $$ l(n)=\lambda(n)+\nu(n)-1 $$
Для x^48 необходимо __6 операций__, для x^63 - __10 операций__ (в шагах 2-5 происходило по 2 операции умножения)

*Теперь применим метод множителей:*
$$ x^{63}\Rightarrow 63=3*21$$
Вычислим x^2 и x^3 - __2 операции__. Пусть y=x^3.
$$ 13 = 1+12 = 1+2*6$$
$$ y^{21}=y^{3^{6}}*y^{3}=y^{3^{2^{3}}} * y^{3} $$
Метод занял 8 операций

$$x^{48}\Rightarrow 48=2*24$$
Вычислим x^2 - __1 операция.__ Пусть y=x^2.
$$y^{24} = y^{8^{3}} = y^{2^{2^{2^{3}}}} = y^{2^{2^{2^{2}}}} * y^{2^{2^{2}}}$$
Метод занял 6 операций.

__Вывод:__
Для числа 48 бинарный метод и метод множителей дали одинаковый результат в 6 операций, для числа 63 бинарный метод занял 10 операций, метод множителей 8 операций.

### Задание 2

Реализовать алгоритм Брауэра (для __нечётных__ вариантов) и алгоритм Яо (для чётных вариантов) для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек
с минимальной аддитивной цепочкой для заданного числа. Сделать
выводы.


In [1]:
def Brauer(n, k):
    
    #список элементов аддитивной цепочки
    answer=[]
    
    #Переведём число в с.с. по основанию 2^k
    lst = n.digits(2^k)
    
    #Добавляем первые элементы цепочки от 1 до 2^k-1
    answer = [i for i in range(1,2^k)]
    
    #Добавим в цепочку старший разряд числа n в с.с. 2^k. Это наше последнее q.
    answer.append(lst[-1])
    
    for i in range(len(lst)-2, -1, -1):
        
        #Запоминаем последний элемент текущей цепочки и "наращиваем его"
        q=answer[-1]
        
        for j in range(1,k+1):
            answer.append((2^j)*q)
            
        #Добавляем к последнему элементу остаток. Теперь в конце цепи мы имеем q "предыдущего уровня".
        answer.append(answer[-1] + lst[i])
        
    #Уберём дубликаты c помощью set() и отсортируем множество с помощью sorted()    
    return sorted(set(answer))

In [2]:
print(Brauer(6969, 3))

[1, 2, 3, 4, 5, 6, 7, 8, 13, 26, 52, 104, 108, 216, 432, 864, 871, 1742, 3484, 6968, 6969]


Сравним длины аддитивных цепочек некоторых чисел, полученных при помощи этого алгоритма, с минимальными цепочками для этих чисел:

In [3]:
print(f"Число 143, k=2 - длина {len(Brauer(143,2))}")
print(f"Число 143, k=3 - длина {len(Brauer(143,3))}")
print(f"Число 133, k=2 - длина {len(Brauer(133,2))}")
print(f"Число 133, k=3 - длина {len(Brauer(133,3))}")
print(f"Число 95, k=2 - длина {len(Brauer(95,2))}")
print(f"Число 95, k=3 - длина {len(Brauer(95,3))}")
print(f"Число 108, k=2 - длина {len(Brauer(108,2))}")
print(f"Число 108, k=3 - длина {len(Brauer(108,3))}")

AttributeError: 'int' object has no attribute 'digits'

| Number | K | Brauer | Min |
|--------|---|--------|-----|
| 143    | 2 | 11     | 11  |
| 143    | 3 | 14     | 11  |
| 133    | 2 | 11     | 10  |
| 133    | 3 | 13     | 10  |
| 95     | 2 | 11     | 10  |
| 95     | 3 | 13     | 10  |
| 108    | 2 | 10     | 9   |
| 108    | 3 | 13     | 9   |

Видно, что при некоторых k алгоритм Брауэра даёт очень близкий к минимальному результат, при других k - значительно более длинный. Оценка алгоритма Брауэра говорит о том, что $$ l(n)\approx\lambda(n)+\frac{\lambda(n)}{\lambda(\lambda(n))}$$


В то время как самая короткая аддитивная цепочка
для числа n имеет длину не более $$ \lambda(n)$$

### Задание 3

Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать,
сколько времени потребовалось на поиск цепочки и какая цепочка
получилась. Сравнить с предыдущими методами, сделать выводы.

Определим несколько полезных функций.

In [7]:
#Длина двоичной записи цепи, уменьшенная на 1
def _lambda(n):
    return len(n.digits(2))-1
    
#Вес Хэмминга
def _nu(n):
    return n.digits(2).count(1)
    
#Нижняя граница оценки длины вектора индексов
def l_floor(n):
    return ceil(log(n,2).n())
    
#Верхняя граница оценки длины вектора индексов
def l_ceil(n):
    return _lambda(n)+_nu(n)-1
    
#Функция, вычисляющая аддитивную цепочку по вектору индексов
def calc_from_index(v):
    answer = []
    answer.append(1)
    for el in v:
        answer.append(answer[-1] + answer[el-1])
    return answer
        
#Проверка на "минимальность" вектора, т. е. состоит ли он из единиц.
def isEmpty(vec):
    for i in vec:
        if i > 1:
            return False
    return True

#Функция, которая уменьшает вектор индексов

def red_vec(vec):
    l = len(vec)
    
    for i in range(l):
        if vec[l-1-i]>1:
            vec[l-1-i] -= 1
            for j in range(l-i,l):
                vec[j]=j+1
            break
            
#Функция генерирует фрагмент вектора индексов заданной длины, начиная с числа first_number
def gen_part(l, first_number):
    part=[]
    for i in range(l):
        part.append(first_number + i)
    return part

Теперь сам алгоритм:

In [8]:
def split_vector(n):
    
    #Вычисляем нижнюю границу m
    m_min=l_floor(n)
    
    #Вычисляем верхнюю границу m
    m_max=l_ceil(n)
    
    #Внешний цикл по длине вектора
    for m in range(m_min, m_max+1):
        
        q = m//2
        #Создаём статичную часть вектора
        static_part = gen_part(q,1)
        
        while True:
            
            #Создаём изменяющуюся часть вектора
            dynamic_part = gen_part(m-q,q+1)
            
            #Вычисляем q+1 - й элемент аддитивной цепочки
            a_q1=calc_from_index(static_part)[-1]
            
            #Найдём минимальное и максимальное возможное число
            a_min = a_q1+m-q
            a_max = a_q1*2^(m-q)
            
            #Если наше число не входит в границы, пропускаем
            if n < a_min or n > a_max:
                if isEmpty(static_part) and len(static_part) > 1:
                    break
                red_vec(static_part)
                continue
                
            #Цикл, перебирающий динамическую часть
            while True:
                
                #Посмотрим, какое число имеем сейчас
                cur_chain = calc_from_index(static_part+dynamic_part)
                
                #Если последнее число в текущей цепочке n, ответ найден
                if cur_chain[-1] == n:
                    return cur_chain
                
                #Если динамическая часть пустая, уменьшать некуда, выход
                if isEmpty(dynamic_part):
                    break
                    
                #Если не пустая - уменьшим
                red_vec(dynamic_part)
             #Не нашли ответ? Проверяем, можно ли уменьшить статичную часть
            if isEmpty(static_part):
                break
            #Уменьшаем статичную часть  
            red_vec(static_part)
            
    #Что-то пошло не так
    return -1

In [9]:
from datetime import datetime
start=datetime.now()
print(split_vector(236))
print(f"Time: {datetime.now() - start}")

[1, 2, 4, 8, 12, 14, 28, 56, 112, 224, 236]
Time: 0:00:00.761859


Протестируем наш алгоритм для нескольких значений больше 1000

In [10]:
from datetime import datetime
numbers = [1019, 1025, 1012, 1128, 1010]
for number in numbers:
    start=datetime.now()
    print(f"Number: {number}")
    print(f"Chain: {split_vector(number)}")
    print(f"Time: {datetime.now() - start}")
    print("--------------------------------")

Number: 1019
Chain: [1, 2, 4, 8, 10, 11, 21, 42, 84, 168, 336, 672, 1008, 1019]
Time: 0:01:11.072627
--------------------------------
Number: 1025
Chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]
Time: 0:00:00.000751
--------------------------------
Number: 1012
Chain: [1, 2, 4, 8, 16, 32, 48, 96, 192, 384, 768, 960, 964, 1012]
Time: 0:00:30.377080
--------------------------------
Number: 1128
Chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1088, 1120, 1128]
Time: 0:00:22.987137
--------------------------------
Number: 1010
Chain: [1, 2, 4, 8, 16, 32, 48, 96, 192, 384, 768, 960, 962, 1010]
Time: 0:00:30.403846
--------------------------------


Попробуем применить Алгоритм Брауэра к этим же числам, используя формулу оптимального k

In [11]:
from datetime import datetime
numbers = [1019, 1025, 1012, 1128, 1010]
for number in numbers:
    start=datetime.now()
    k = 2
    print(f"Number: {number}, k: {k}")
    print(f"Chain: {Brauer(number, k)}")
    print(f"Time: {datetime.now() - start}")
    print("--------------------------------")

Number: 1019, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 254, 508, 1016, 1019]
Time: 0:00:00.000616
--------------------------------
Number: 1025, k: 2
Chain: [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]
Time: 0:00:00.000064
--------------------------------
Number: 1012, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 253, 506, 1012]
Time: 0:00:00.000109
--------------------------------
Number: 1128, k: 2
Chain: [1, 2, 3, 4, 8, 16, 17, 34, 68, 70, 140, 280, 282, 564, 1128]
Time: 0:00:00.000181
--------------------------------
Number: 1010, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 504, 1008, 1010]
Time: 0:00:00.000588
--------------------------------


Number: 1019, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 254, 508, 1016, 1019]
Time: 0:00:00.000553
--------------------------------
Number: 1025, k: 2
Chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]
Time: 0:00:00.000293
--------------------------------
Number: 1012, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 253, 506, 1012]
Time: 0:00:00.000206
--------------------------------
Number: 1128, k: 2
Chain: [1, 2, 4, 8, 16, 17, 34, 68, 70, 140, 280, 282, 564, 1128]
Time: 0:00:00.000275
--------------------------------
Number: 1010, k: 2
Chain: [1, 2, 3, 6, 12, 15, 30, 60, 63, 126, 252, 504, 1008, 1010]
Time: 0:00:00.000203
--------------------------------

| Number | Brauer | Vector crush |
|--------|--------|--------------|
| 1019   | 15     | 14           |
| 1025   | 12     | 12           |
| 1012   | 14     | 14           |
| 1128   | 14     | 14           |
| 1010   | 14     | 14           |

__Вывод:__ Алгоритм был протестирован для тех же чисел, что и алгоритм Брауэра. Длины звёздных цепочек почти всегда такие же, как и для алгоритма Брауэра. Алгоритм дробления вектора индексов работает __очень долго__, т.к. является перебором, хоть и несколько улучшенным. Время выполнения алгоритма дробления вектора может занимать минуты и часы. Также было замечено, что быстрее всего он работает для чисел, которые чуть больше степеней двойки, а хуже всего для чисел, чуть меньше степеней двойки.

### Задание 4

 Проверить гипотезу Шольца–Брауэра для всех натуральных чисел от 1 до 12 на алгоритме дробления вектора индексов. Результаты
оформить в виде таблицы. Сделать выводы.

In [ ]:
for n in range(1,12 + 1):
    a = Integer(n)
    b = Integer(2**n-1)
    chain_a = split_vector(a)
    chain_b = split_vector(b)
    len_a = len(chain_a)
    len_b = len(chain_b)
    print(f"n={n}, l(n) = {len_a}, l(2^n-1) = {len_b}")
    print(f"l(2^n - 1) <= l(n) + n - 1: {len_b <= n - 1 + len_a}")
    print()

n=1, l(n) = 1, l(2^n-1) = 1
l(2^n - 1) <= l(n) + n - 1: True

n=2, l(n) = 2, l(2^n-1) = 3
l(2^n - 1) <= l(n) + n - 1: True

n=3, l(n) = 3, l(2^n-1) = 5
l(2^n - 1) <= l(n) + n - 1: True

n=4, l(n) = 3, l(2^n-1) = 6
l(2^n - 1) <= l(n) + n - 1: True

n=5, l(n) = 4, l(2^n-1) = 8
l(2^n - 1) <= l(n) + n - 1: True

n=6, l(n) = 4, l(2^n-1) = 9
l(2^n - 1) <= l(n) + n - 1: True

n=7, l(n) = 5, l(2^n-1) = 11
l(2^n - 1) <= l(n) + n - 1: True

n=8, l(n) = 4, l(2^n-1) = 11
l(2^n - 1) <= l(n) + n - 1: True

n=9, l(n) = 5, l(2^n-1) = 13
l(2^n - 1) <= l(n) + n - 1: True



n=1, l(n) = 1, l(2^n-1) = 1
l(2^n - 1) <= l(n) + n - 1: True

n=2, l(n) = 2, l(2^n-1) = 3
l(2^n - 1) <= l(n) + n - 1: True

n=3, l(n) = 3, l(2^n-1) = 5
l(2^n - 1) <= l(n) + n - 1: True

n=4, l(n) = 3, l(2^n-1) = 6
l(2^n - 1) <= l(n) + n - 1: True

n=5, l(n) = 4, l(2^n-1) = 8
l(2^n - 1) <= l(n) + n - 1: True

n=6, l(n) = 4, l(2^n-1) = 9
l(2^n - 1) <= l(n) + n - 1: True

n=7, l(n) = 5, l(2^n-1) = 11
l(2^n - 1) <= l(n) + n - 1: True

n=8, l(n) = 4, l(2^n-1) = 11
l(2^n - 1) <= l(n) + n - 1: True

n=9, l(n) = 5, l(2^n-1) = 13
l(2^n - 1) <= l(n) + n - 1: True

n=10, l(n) = 5, l(2^n-1) = 14
l(2^n - 1) <= l(n) + n - 1: True

n=11, l(n) = 6, l(2^n-1) = 16
l(2^n - 1) <= l(n) + n - 1: True

n=12, l(n) = 5, l(2^n-1) = 16
l(2^n - 1) <= l(n) + n - 1: True


Результаты в виде таблицы:

| n  | l(n) | l(2^n-1) |
|----|------|----------|
| 1  | 1    | 1        |
| 2  | 2    | 3        |
| 3  | 3    | 5        |
| 4  | 3    | 6        |
| 5  | 4    | 8        |
| 6  | 4    | 9        |
| 7  | 5    | 11       |
| 8  | 4    | 11       |
| 9  | 5    | 13       |
| 10 | 5    | 14       |
| 11 | 6    | 16       |
| 12 | 5    | 16       |

__Вывод:__
Гипотеза подтверждена для n от 1 до 12.

## Выводы

Общий вывод по проделанной работе.

__Вывод:__ В ходе выполнения данной практической работы были сформированы представления об аддитивных цепочках, были выработаны умения реализовывать и применять алгоритмы для нахождения минимальных аддитивных цепочек, получеен навык использования системы компьютерной математики SageMath для реализации таких алгоритмов. На практике была проверена гипотеза Шольца-Брауэра, были проанализированы алгоритм Брауэра и алгоритм дробления векторов индексов.